### PR TITLE
A couple of small refactorings

### PR DESCRIPTION
**Pull request by nicolaiparlog**
I hereby agree to the terms of the JUnit Contributor License Agreement.

### CODE REVIEW

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java [hunk 1]
```diff
@@ -88,28 +90,47 @@ private ExtensionRegistry(ExtensionRegistry parent) {
...
 	 * @see #getReversedExtensions(Class)
-	 * @see #stream(Class)
+	 * @see #getExtensions(Class)
 	 */
-	public <E extends Extension> List<E> getExtensions(Class<E> extensionType) {
-		List<E> extensions = new ArrayList<>();
-
-		if (this.parent != null) {
-			extensions.addAll(this.parent.getExtensions(extensionType));
+	public <E extends Extension> Stream<E> streamExtensions(Class<E> extensionType) {
```
Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: Wouldn't this read better as a noun rather than a verb? E.g. `extensionsStream` rather than `streamExtensions`?

Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: Or, if you want more consistency with other methods like `getExtensions`, how about `getExtensionsAsStream`? (I'm less eager to suggest this on though, as I think `extensionsStream` is pleasingly shorter and easier to read.)

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java [hunk 2]
```diff
@@ -88,28 +90,47 @@ private ExtensionRegistry(ExtensionRegistry parent) {
...
 		}
+	}
 
+	/**
+	 * Stream all {@code Extensions} in this registry (ancestor are ignored!).
+	 *
+	 * @param extensionType the type of {@link Extension} to get
+	 * @see #getReversedExtensions(Class)
+	 */
+	private <E extends Extension> Stream<E> streamLocalExtensions(Class<E> extensionType) {
```
Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: See above.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java (hunk 1 above)
Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: But then again, it seems to fit JUnit 5's code style more than `extensionsStream`, so I suppose I wouldn't mind either way.

Comment (nicolaiparlog) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: I actually prefer verbs for methods/functions - I'd maybe use nouns for properties if I ditch the get/set. It could be argued that this is such a case. But as you guessed, I wanted to keep this close to `getExtensions` and `getExtensionsAsStream` does that even better so I'm open to that as well. :smiley:

Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: Okay, sounds good to me. :smiley:

Just so that you know where I'm coming from, the reason I disliked the idea of `streamExtensions` was because in OpenJDK 9 there is a forthcoming method `JarFile::versionedStream` (https://bugs.openjdk.java.net/browse/JDK-8163798) which to me looks and feels more like a noun than a verb and which uses "Stream" at the end of the phrase rather than the beginning, so I thought that, if you wanted to stick with something like `streamExtensions`, it would be less jarring and more consistent to use `extensionsStream`. But now that we've both decided on `getExtensionsAsStream`, which follows JUnit 5's coding style better, this is no longer a concern for me. :smile:

Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: (Oops, I was a bit quick to say that we've "decided" on `getExtensionsAsStream`, as I've realised we haven't made a firm decision yet! So I suppose I just want to say that, if you're happy with `getExtensionsAsStream`, then I'm happy that it as well!)

**Comment by sbrannen** [comment]
@nicolaiparlog, regarding the "Refactor extension accessors in 'ExtensionFactory'" commit, please change 'ExtensionFactory' to ExtensionRegistry.

And when you say "no delegation", I'm assuming you mean "now delegation...".

Also, in line with our [contributing guidelines](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#commit-messages), please wrap commit comments around 74 characters.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java (hunk 1 above)
Comment (sbrannen) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: Hi guys,

While I certainly appreciate your attention to detail, let's not go overboard here. This is _just_ an internal API. 😉 

The reason the method is currently called `stream` is for DSL-like conciseness. 

Although I can live with `streamExtensions`, I will personally veto `getExtensionsAsStream` since there is no need to be overly verbose here.

The _thing_ in question is an extension registry. Thus getting a stream from it will almost certainly _stream_ the extensions contained in that registry. Honestly, what else would you expect such a method to do? Plus, the method clearly documents what it does.

In summary, I welcome the other changes in this PR, but I don't see a concrete need for renaming the `stream` method.

@junit-team/junit-lambda, what do the rest of you think?

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java [hunk 3]
```diff
@@ -159,6 +159,11 @@ public boolean isSkipped() {
 		public Optional<String> getReason() {
 			return Optional.ofNullable(this.reason);
 		}
+
+		@Override
+		public String toString() {
+			return skipped ? String.format("Skip [%s]", reason) : "Do not skip";
+		}
```
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java]: This is not how we implement `toString()`.

Please rework in line with `org.junit.jupiter.api.extension.ConditionEvaluationResult.toString()`.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java (hunk 1 above)
Comment (jbduncan) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: > While I certainly appreciate your attention to detail, let's not go overboard here. This is just an internal API. :wink:

Haha, fair enough!

> Although I can live with streamExtensions, I will personally veto getExtensionsAsStream since there is no need to be overly verbose here.

Okay then. I can see that I was overly bikeshedding and letting myself be OCD here, so I'm happy for `streamExtensions` to be as-is. :smiley:

Comment (nicolaiparlog) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: > The _thing_ in question is an extension registry. Thus getting a stream from it will almost certainly stream the extensions contained in that registry.

Yes but the same argument could be made for `getExtensions`. I was just irked by the fact that their names differed more than their intent and think they should be unified. But whether it's `get`/`stream`, `getExtensions`/`streamExtensions`, or `getExtensions`/`getExtensionsAsStream` is a secondary concern (although I like the PR's variant best).

But let's not get distracted by bikeshedding. The much more "important" (well, relatively) change is that delegation order changed. I guess that part is ok?

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java (hunk 3 above)
Comment (nicolaiparlog) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java]: Will do. Quick question on JUnit's stance on `Optional` fields - do you still do that? `ConditionEvaluationResult` uses one (`reason`) and `SkipResult` might for the same, err, `reason`. Personally I like that style - purposely ignoring the JDK team's recommendations - but I wanted to make sure that this is not just an old artifact.

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java (hunk 1 above)
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: > In summary, I welcome the other changes in this PR, but I don't see a concrete need for renaming the `stream` method.

👍

**Comment by marcphilipp** [comment]
Team decisions:

- Please re-rename `streamExtensions` to `stream`
- Changing the delegation order is fine
- Please change Node.SkipResult to store an `Optional` to be consistent with `ConditionEvaluationResult`

File: junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java [hunk 4]
```diff
@@ -88,28 +90,46 @@ private ExtensionRegistry(ExtensionRegistry parent) {
...
 	 * @see #getReversedExtensions(Class)
-	 * @see #stream(Class)
+	 * @see #getExtensions(Class)
 	 */
-	public <E extends Extension> List<E> getExtensions(Class<E> extensionType) {
-		List<E> extensions = new ArrayList<>();
-
-		if (this.parent != null) {
-			extensions.addAll(this.parent.getExtensions(extensionType));
+	public <E extends Extension> Stream<E> streamExtensions(Class<E> extensionType) {
```
Comment (marcphilipp) [inline junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/extension/ExtensionRegistry.java]: Please re-rename `streamExtensions` to `stream`.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java [hunk 5]
```diff
@@ -159,6 +159,11 @@ public boolean isSkipped() {
 		public Optional<String> getReason() {
 			return Optional.ofNullable(this.reason);
```
Comment (marcphilipp) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java]: Please store an `Optional` like we do in `ConditionEvaluationResult`.

File: junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java (hunk 3 above)
Comment (sbrannen) [inline junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/Node.java]: Nope. That's not an old artifact. 

We _still do that_ if we feel it's appropriate.

**Comment by nicolaiparlog** [comment]
All done. I still think that naming them `getExtensions` and `stream` is inconsistent but I'll capitulate. :crying_cat_face:

**Comment by sbrannen** [comment]
@marcphilipp, can you take care of the "1 review requesting changes"?

**Comment by sbrannen** [comment]
Thanks for the PR, @nicolaiparlog! 👍 

It has been merged into `master`.
